### PR TITLE
Ce/parallel agg local

### PR DESCRIPTION
**Pull request by calellowitz**
@snopoke 
Is this what you were going for? I wasn't familiar with `thread_local` before this, but it does seem to make this WAY simpler. This does rely on the citus db having the alias `citus` but otherwise I do not think i made any core decisions.

### CODE REVIEW

File: corehq/sql_db/routers.py [hunk 1]
```diff
@@ -102,9 +107,12 @@ def db_for_read_write(model, write=True):
...
-        engine_id = ICDS_UCR_ENGINE_ID
-        if not write:
-            engine_id = connection_manager.get_load_balanced_read_db_alias(ICDS_UCR_ENGINE_ID)
+        if getattr(_thread_local, 'force_citus', False):
+            engine_id = CITUS_MASTER
+        else:
+            engine_id = ICDS_UCR_ENGINE_ID
+            if not write:
+                engine_id = connection_manager.get_load_balanced_read_db_alias(ICDS_UCR_ENGINE_ID)
         return connection_manager.get_django_db_alias(engine_id)
```
Comment (sravfeyn) [inline corehq/sql_db/routers.py]: This wat, doesn't the other uninteded ICDS models also get routed to Citus, like `ICDSAuditEntryRecord` etc for which we are not doing citus writes?

File: corehq/sql_db/routers.py [hunk 2]
```diff
@@ -102,9 +107,12 @@ def db_for_read_write(model, write=True):
...
     elif app_label == ICDS_REPORTS_APP:
-        engine_id = ICDS_UCR_ENGINE_ID
-        if not write:
-            engine_id = connection_manager.get_load_balanced_read_db_alias(ICDS_UCR_ENGINE_ID)
+        if getattr(_thread_local, 'force_citus', False):
+            engine_id = CITUS_MASTER
+        else:
+            engine_id = ICDS_UCR_ENGINE_ID
+            if not write:
+                engine_id = connection_manager.get_load_balanced_read_db_alias(ICDS_UCR_ENGINE_ID)
```
Comment (sravfeyn) [inline corehq/sql_db/routers.py]: This way, doesn't the other uninteded ICDS models also get routed to Citus, like `ICDSAuditEntryRecord` etc for which we are not doing citus writes?

Comment (calellowitz) [reply corehq/sql_db/routers.py]: This is only used in the aggregation, and I do not think those models are used in the aggregation, but if they are we want to make sure they still work

**Comment by calellowitz** [comment]
@snopoke 
I think this is now complete, but would love a final pass before we merge

File: custom/icds_reports/tasks.py [hunk 3]
```diff
@@ -174,10 +176,16 @@
...
+    if toggles.PARALLEL_AGGREGATION.enabled(DASHBOARD_DOMAIN):
+        move_ucr_data_into_aggregation_tables.delay(force_citus=True)
 
 
 @serial_task('move-ucr-data-into-aggregate-tables', timeout=36 * 60 * 60, queue='icds_aggregation_queue')
-def move_ucr_data_into_aggregation_tables(date=None, intervals=2):
+def move_ucr_data_into_aggregation_tables(date=None, intervals=2, force_citus=False):
+
+    if force_citus:
+        force_citus_engine()
```
Comment (snopoke) [inline custom/icds_reports/tasks.py]: should this be unset afterwards? Might be safer to make a context manager that reset's the value on exit

Comment (snopoke) [reply custom/icds_reports/tasks.py]: This thread local won't get applied to the sub-tasks that get executed. I think you'll still have to pass `force_citus` to the sub-tasks.

**Review by snopoke — CHANGES_REQUESTED** [review]

**Review by snopoke — APPROVED** [review]

**Pull request merged**